### ISSUE TITLE
Reduce dependencies on third party npm modules

### ISSUE THREAD

**Issue by mereacre** [nqminds/nqm-iot-database-utils#36]
We need to reduce the dependencies on third party npm modules in order to make the code as clean as possible and to use latest Javascript functionalities.

We have the following modules:
    "bluebird": "^3.5.0",
    "del": "^3.0.0",
    "lodash": "^4.17.15",
    "mongo-sql": "^5.0.4",
    "path": "^0.12.7",
    "shortid": "^2.2.8",
    "sqlite3": "^4.0.4",
    "temp-dir": "^1.0.0"

We need to check which module can be replaced with basic Javascript functionality.

We should start with lodash.

**Comment by IGlautier** [comment]
I've taken a quick look at the list above and notes are below:

- path is core node
- shortid has 1 dependency which has 0 dependencies
- lodash has 0 dependencies and is a well known library and offers some really clean helper functions. While I agree you should use native forEach, map, and reduce where possible in some cases lodash results in much cleaner code.
- We can almost certainly replace bluebird with native promise unless you need functionality like Promise.delay
- del has a fairly big dependency tree given the functionality it appears to expose: https://npm.anvaka.com/#/view/2d/del @aloisklink maybe you can replicate the functionality you need using fs module?
- mongo-sql has 0 dependencies
- temp-dir has 0 dependencies but the functionality looks very simple, maybe could roll your own
- sqlite3 has loads of dependencies https://npm.anvaka.com/#/view/2d/sqlite3
However the functionality is too complex to reimplement

@mereacre What's the priority on this? I feel like trying to cull the already quite small dependency tree too early might be a mistake.

**Comment by mereacre** [comment]
It has no priority. It is a guideline. I think we should definitely replace eventually lodash and others. A lot of functions are in native JavaScript. Also we reduce the security vulnerabilities possibly introduced in third party projects.

**Comment by IGlautier** [comment]
I agree on reducing use of lodash and others when there is exact functionality offered by native. But I don't think we should push to avoid using a well developed library when it has a solution that we need. Any security issues or bugs in lodash are likely to be discovered and patched quite fast versus anything we build internally and we just don't have the man power to replicate the functionality of more complex libraries.

**Comment by aloisklink** [comment]
- I believe the only thing of Bluebird we use is [`Promse.promisifyAll`](http://bluebirdjs.com/docs/api/promise.promisifyall.html) for the `sqlite3` library. If we switch to the [`sqlite`](https://www.npmjs.com/package/sqlite) we can get easily get rid of it, but we'll add a less used library.
- `del` should be easy to remove, it is only used to delete one folder.
- `temp-dir` should definitely be removed. It looks like it is literally just `fs.realpathSync(os.tmpdir())` to replace it.
- `path` looks like it should only be used for non-`Node.js` code so we can remove it.

**Comment by mereacre** [comment]
I agree about not writing our code when there's lots of stuff already in the npm repos. When I wrote the first version of database utils I used lots of packages to save development time. But looking at the code now and what Alois is saying it makes perfect sense to remove the packages that do not add any significant value to the project. I would even consider extracting the json to SQL part from mongo-sql package as we are using only a very small subset of what mongo-sql is offering. But for know I think we should keep it as it is. In general I think we should aim for "pure" JavaScript packages with code that we can understand, debug and test properly.

**Comment by TobyEalden** [comment]
This is completely bonkers!

**Comment by IGlautier** [comment]
Tactfully put
